trace_tcp: Add fd (updated from Burak's #3850)

alban opened the pull request:
# trace_tcp: Add fds for connect and accept

Supersedes #3850.
Part of #3565

The fields are documented in gadget.yaml:

- The 'fd' field is the fd where the operation happened on (connect, accept)
- The 'accept_fd' field is the FD of the new connection returned by accept, or -1 for errors cases or other events

The gadget uses maps to store context. Map cleanup is always done in a tracepoint or a kprobe, avoiding kretprobe. This ensures it does not miss a cleanup because the kretprobe overflows the maxactive.

TCP close events don't include a fd because this is not always well-defined: a fd can be "renamed" with dup2() so it cannot be matched with the fd from connect or accept events.

## How to use

```bash
$ nc -k -l -p 8080 &
[1] 588902
$ echo|nc 127.0.0.1 8080
```

```bash
$ sudo ./ig run ghcr.io/inspektor-gadget/gadget/trace_tcp:alban_trace_tcp_fd2 --verify-image=false --host --fields proc.comm,proc.pid,proc.tid,src,dst,type,fd,accept_fd,error
WARN[0001] image signature verification is disabled due to using corresponding option
WARN[0001] image signature verification is disabled due to using corresponding option
COMM     PID     TID SRC               DST               TYPE     FD  ACCEPT_FD ERROR
nc    588910  588910 127.0.0.1:49380   127.0.0.1:8080    connect  3   -1
nc    588902  588902 127.0.0.1:8080    127.0.0.1:49380   accept   4   5
nc    588902  588902 127.0.0.1:8080    127.0.0.1:49380   close    -1  -1
nc    588910  588910 127.0.0.1:49380   127.0.0.1:8080    close    -1  -1
```

## Testing done

The CI is green

mauriciovasquezbernal commented on gadgets/trace_tcp/ebpf/connect.h:
```diff
@@ -53,36 +64,50 @@ static __always_inline int exit_tcp_connect(struct pt_regs *ctx, int ret,
 	if (!skpp)
 		return 0;
 
+	fd = bpf_map_lookup_elem(&tcp_tid_fd, &tid);
+	if (!fd)
+		return 0;
+	ei.fd = *fd;
+
 	if (ret)
```
Not fully related to this PR, but can't we move the `if (ret)` to be the first one to avoid wasting cpu cycles on the map lookups in case ret != 0?

mauriciovasquezbernal reviewed (commented):
I have some comments. I haven't touch this gadget for a while, so I'm a little bit lost. My only blocking comment so far is about tcp_sock_fd not being used. I'll try to make a deeper pass tomorrow.

alban commented on gadgets/trace_tcp/ebpf/common.h:
```diff
@@ -44,6 +46,68 @@ struct tuple_key_t {
 	u32 netns;
 };
 
+/*
+ * For tcp_close it can happen that it is not called through sys_close
+ * This is a best effort approach to still provide the fd through other
+ * kprobes.
+ */
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct sock *);
+	__type(value, __u32); // fd
+} tcp_sock_fd SEC(".maps");
```
You're right. I forgot to clean it up in my commit 548d5049236b864ddf8c37e8c2c59d977d622c2b.

Burak initially implemented the fds for all events: connect, accept and close. I removed the fds for the close event because:
- it is not clearly defined (sometimes it gives the fd given as parameter to the close syscall, sometimes the fd number that was used in the connect syscall before it was renamed via dup2())
- it does not always work (if ig is started after the connect event and if the tcp-close event is done outside of the close syscall)
- users didn't request for it. I've heard user requests for fds in connect and accept events but not for close events, so we don't have to figure out everything in this PR.

alban commented:
This CI failure seems unrelated to the PR:
```
time="2025-01-22T16:46:19Z" level=warning msg="getParamValue: param \"non-existing-param\" not found"
time="2025-01-22T16:46:19Z" level=warning msg="getSyscallDeclaration: getting syscall declaration for \"foobar\": no syscall correspond to \"foobar\""
time="2025-01-22T16:46:19Z" level=error msg="instantiating operator \"wasm\": initializing wasm: instantiating wasm: import func[env.getSyscallDeclaration]: signature mismatch: i64_i32 != i64i64_i32"
time="2025-01-22T16:46:19Z" level=warning msg="perfReaderRead: reading perf buffer: perf ringbuffer: must have been paused before reading overwritable buffer"
time="2025-01-22T16:46:19Z" level=warning msg="perfReaderRead: reading perf buffer: epoll wait: i/o timeout"
time="2025-01-22T16:46:19Z" level=error msg="reading perf record"
=== NAME  TestPerf
    wasm_test.go:661: 
        	Error Trace:	/home/runner/work/inspektor-gadget/inspektor-gadget/pkg/operators/wasm/wasm_test.go:661
        	Error:      	Received unexpected error:
        	            	starting operators: starting operator "oci": starting operator "wasm": gadgetStart failed
        	Test:       	TestPerf
        	Messages:   	running gadget
--- FAIL: TestPerf (0.83s)
```

I suspect it is related to #3725. cc @eiffel-fl 

I'll just restart the CI.

mauriciovasquezbernal commented:
> I suspect it is related to #3725. cc @eiffel-fl
> 

Yes, I believe that test is flaky. I've seen the same on other runs.

mauriciovasquezbernal reviewed (approved):
I did some manual tests and it works as expected. My only comment is on the new_fd name, perhaps accept_fd is more clear? I don't have a strong opinion, so feel free to update or just merge this PR.

Thanks!

alban commented on gadgets/trace_tcp/ebpf/accept.h:
```diff
@@ -9,34 +9,74 @@
 
-	if (!sk)
-		return;
+	fdPtr = bpf_map_lookup_elem(&tcp_tid_fd, &tid);
```
clang-format does not complain. But I can change it to fd_ptr.

alban commented on gadgets/trace_tcp/ebpf/accept.h:
```diff
@@ -9,34 +9,74 @@
+		goto end;
 
 	fill_event(event, &t, NULL, 0, accept);
+	event->fd = fd;
+
+	if (ctx->ret >= 0)
+		event->new_fd = (int)ctx->ret;
+	else
+		event->new_fd = -1;
```
Good idea. The compiler will optimize it out if overwritten after.

alban commented:
> the new_fd name, perhaps accept_fd is more clear

It makes sense. I renamed it.

> What is the goal of this PR? To merge in main or in Burak's branch?
> If goal is to merge in main, the history must be cleaned with "fixup" commits squashed.

I want to merge it into main. I squashed the commits and I will address your remaining comments before merging.
